File: packages/typescript-sdk/src/prompts.ts
/**
 * Prompt registry fetch helper (PRD-4).
 *
 * Resolve a managed prompt by name + label (Langfuse-style) at runtime — so you
 * can ship prompt changes by promoting a label, with no code redeploy. Results
 * are cached in-memory by name+label with a short TTL; promote a label and the
 * next fetch past the TTL picks it up.
 *
 *   import { getPrompt } from "@prism-llm-labs/sdk";
 *   const p = await getPrompt("support-reply", { label: "production" });
 *   const messages = p.compile({ customer: "Dana" });   // fills {{customer}}
 *   // pass p.promptVersion as tags['prompt_version'] so spend/quality attribute to it
 *
 * Server: GET /api/prompts/resolve (authenticated by PRISM_API_KEY).
 */
export interface PromptMessage { role: string; content: string }

export interface ResolvedPrompt {
  name:          string;
  version:       number;
  messages:      PromptMessage[];
  config:        Record<string, unknown>;
  /** "name@version" — stamp as tags['prompt_version'] so attribution flows. */
  promptVersion: string;
  /** Fill {{variable}} placeholders in message contents; returns a new messages array. */
  compile:       (variables?: Record<string, string | number>) => PromptMessage[];
}

export interface GetPromptOptions {
  label?:     string;     // default "production"
  version?:   number;     // pin an explicit version (overrides label)
  projectId?: string;
  apiKey?:    string;     // default process.env.PRISM_API_KEY
  baseUrl?:   string;
  /** Cache TTL in ms (default 60000). Pass 0 to bypass the cache. */
  ttlMs?:     number;
}

interface CacheEntry { value: ResolvedPrompt; expiresAt: number }
const _cache = new Map<string, CacheEntry>();

function resolveBaseUrl(explicit?: string): string {
  const url =
    explicit ??
    process.env["PRISM_GATEWAY_URL"] ??
    process.env["PRISM_APP_URL"] ??
    process.env["NEXT_PUBLIC_APP_URL"] ??
    "https://useprism.dev";
  return url.replace(/\/$/, "");
}

function compileMessages(messages: PromptMessage[], variables?: Record<string, string | number>): PromptMessage[] {
  if (!variables) return messages.map(m => ({ ...m }));
  return messages.map(m => ({
    role:    m.role,
    content: m.content.replace(/\{\{\s*(\w+)\s*\}\}/g, (_, k) => (k in variables ? String(variables[k]) : `{{${k}}}`)),
  }));
}

/** Clear the in-memory prompt cache (e.g. in tests, or after a known promote). */
export function clearPromptCache(): void { _cache.clear(); }

export async function getPrompt(name: string, opts: GetPromptOptions = {}): Promise<ResolvedPrompt> {
  const apiKey = opts.apiKey ?? process.env["PRISM_API_KEY"];
  if (!apiKey) throw new Error("Prism getPrompt: missing API key (set PRISM_API_KEY or pass apiKey).");

  const label = opts.label ?? "production";
  const ttl   = opts.ttlMs ?? 60_000;
  const key   = `${name}:${opts.version != null ? `v${opts.version}` : label}:${opts.projectId ?? ""}`;

  if (ttl > 0) {
    const hit = _cache.get(key);
    if (hit && hit.expiresAt > Date.now()) return hit.value;
  }

  const qs = new URLSearchParams({ name });
  if (opts.version != null) qs.set("version", String(opts.version));
  else qs.set("label", label);
  if (opts.projectId) qs.set("project_id", opts.projectId);

  const res = await fetch(`${resolveBaseUrl(opts.baseUrl)}/api/prompts/resolve?${qs.toString()}`, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  const json = await res.json().catch(() => ({})) as Record<string, unknown>;
  if (!res.ok) throw new Error(`Prism getPrompt failed (${res.status}): ${String(json.error ?? res.statusText)}`);

  const messages = (Array.isArray(json.content) ? json.content : []) as PromptMessage[];
  const value: ResolvedPrompt = {
    name:          String(json.name ?? name),
    version:       Number(json.version ?? 0),
    messages,
    config:        (json.config as Record<string, unknown>) ?? {},
    promptVersion: String(json.prompt_version ?? `${name}@${json.version ?? 0}`),
    compile:       (variables) => compileMessages(messages, variables),
  };

  if (ttl > 0) _cache.set(key, { value, expiresAt: Date.now() + ttl });
  return value;
}

File: packages/python-sdk/prism/prompts.py
"""
Prompt registry fetch helper (PRD-4).

Resolve a managed prompt by name + label (Langfuse-style) at runtime — ship
prompt changes by promoting a label, no code redeploy. Results are cached
in-memory by name+label with a short TTL.

    from prism.prompts import get_prompt
    p = get_prompt("support-reply", label="production")
    messages = p.compile({"customer": "Dana"})   # fills {{customer}}
    # pass p.prompt_version as tags['prompt_version'] so attribution flows

Server: GET /api/prompts/resolve (authenticated by PRISM_API_KEY).
"""
from __future__ import annotations

import os
import re
import time
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

import httpx

_DEFAULT_APP_URL = "https://prism-dip-dey-s-projects.vercel.app"
_VAR_RE = re.compile(r"\{\{\s*(\w+)\s*\}\}")

# key -> (value, expires_at_epoch_seconds)
_cache: Dict[str, Tuple["ResolvedPrompt", float]] = {}


@dataclass
class ResolvedPrompt:
    name: str
    version: int
    messages: List[Dict[str, str]]
    config: Dict[str, Any]
    prompt_version: str  # "name@version" → stamp as tags['prompt_version']

    def compile(self, variables: Optional[Dict[str, Any]] = None) -> List[Dict[str, str]]:
        """Fill {{variable}} placeholders; returns a new messages list."""
        if not variables:
            return [dict(m) for m in self.messages]

        def repl(match: "re.Match[str]") -> str:
            key = match.group(1)
            return str(variables[key]) if key in variables else match.group(0)

        return [{"role": m["role"], "content": _VAR_RE.sub(repl, m["content"])} for m in self.messages]


def clear_prompt_cache() -> None:
    _cache.clear()


def _base_url(explicit: Optional[str]) -> str:
    url = explicit or os.environ.get("PRISM_GATEWAY_URL") or os.environ.get("PRISM_APP_URL") or _DEFAULT_APP_URL
    return url.rstrip("/")


def get_prompt(
    name: str,
    *,
    label: Optional[str] = None,
    version: Optional[int] = None,
    project_id: Optional[str] = None,
    api_key: Optional[str] = None,
    base_url: Optional[str] = None,
    ttl_ms: int = 60_000,
    timeout: float = 10.0,
) -> ResolvedPrompt:
    key = api_key or os.environ.get("PRISM_API_KEY")
    if not key:
        raise ValueError("Prism get_prompt: missing API key (set PRISM_API_KEY or pass api_key).")

    label_used = label or "production"
    cache_key = f"{name}:{'v' + str(version) if version is not None else label_used}:{project_id or ''}"
    if ttl_ms > 0:
        hit = _cache.get(cache_key)
        if hit and hit[1] > time.time():
            return hit[0]

    params: Dict[str, str] = {"name": name}
    if version is not None:
        params["version"] = str(version)
    else:
        params["label"] = label_used
    if project_id:
        params["project_id"] = project_id

    resp = httpx.get(
        f"{_base_url(base_url)}/api/prompts/resolve",
        params=params,
        headers={"Authorization": f"Bearer {key}"},
        timeout=timeout,
    )
    try:
        data = resp.json()
    except Exception:
        data = {}
    if resp.status_code >= 300:
        raise RuntimeError(f"Prism get_prompt failed ({resp.status_code}): {data.get('error', resp.text)}")

    content = data.get("content") or []
    value = ResolvedPrompt(
        name=str(data.get("name", name)),
        version=int(data.get("version", 0)),
        messages=[{"role": m.get("role", ""), "content": m.get("content", "")} for m in content],
        config=data.get("config") or {},
        prompt_version=str(data.get("prompt_version", f"{name}@{data.get('version', 0)}")),
    )
    if ttl_ms > 0:
        _cache[cache_key] = (value, time.time() + ttl_ms / 1000.0)
    return value
